### PR TITLE
Only use StatusCode values in Status.

### PR DESCRIPTION
**Pull request by coryan**
Remove the constructors that accept `long`, and only use the values in
the StatusCode enum. We use the `storage::AsStatus()` function to map
the HttpResponse to the correct `Status`.

Lots of changes in tests because instead of checking for HTTP codes we
need to check for the corresponding StatusCode.

The more subtle change is what happens with `IsRetryable`, I kept the
semantics of its tests unchanged, so the behavior should be the same.

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/google-cloud-cpp/1826)

### CODE REVIEW

File: google/cloud/storage/tests/curl_resumable_upload_session_integration_test.cc [hunk 1]
```diff
@@ -82,7 +82,7 @@ TEST_F(CurlResumableUploadIntegrationTest, WithReset) {
   std::string const contents(UploadChunkRequest::kChunkSizeQuantum, '0');
   StatusOr<ResumableUploadResponse> response =
       (*session)->UploadChunk(contents, 2 * contents.size());
-  ASSERT_TRUE(response.status().status_code() == 200 or response.status().status_code() == 308)
+  ASSERT_TRUE(response.status().ok())
```
Comment (devbww) [inline google/cloud/storage/tests/curl_resumable_upload_session_integration_test.cc]: Join lines?

Comment (coryan) [reply google/cloud/storage/tests/curl_resumable_upload_session_integration_test.cc]: Why, yes.  I applied clang-format manually, I do not understand why it passed the clang-format checks.

File: google/cloud/storage/internal/curl_client_test.cc [hunk 2]
```diff
@@ -33,7 +33,7 @@ namespace {
 using ::google::cloud::storage::oauth2::Credentials;
 using ::testing::HasSubstr;
 
-long const STATUS_ERROR_CODE = 503;
+StatusCode STATUS_ERROR_CODE = StatusCode::kUnavailable;
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/internal/curl_client_test.cc]: Yes. Fixed.

**Comment by coryan** [comment]
The CI test on Kokoro+Ubuntu broke due to #1829.  I will send a separate PR for that and merge this PR only after #1829 is fixed.

**Comment by coryan** [comment]
Rebased to pick up the 308 handling fixes.

**Pull request merged**